File: core/src/main/java/heroes/journey/utils/worldgen/MapGenerationEffect.java
package heroes.journey.utils.worldgen;

import heroes.journey.GameState;
import lombok.Getter;
import lombok.NonNull;

import java.util.ArrayList;
import java.util.List;

import static heroes.journey.registries.Registries.MapGenerationManager;

@Getter
public abstract class MapGenerationEffect {
    @NonNull
    private final String name;
    public final List<String> dependsOn;

    public MapGenerationEffect(String name) {
        this.name = name;
        this.dependsOn = new ArrayList<>();
    }

    public MapGenerationEffect register(MapGenerationEffect... dependencies) {
        MapGenerationManager.register(this);
        for (MapGenerationEffect dependency : dependencies) {
            dependsOn.add(dependency.name);
        }
        return this;
    }

    public abstract void apply(GameState gameState);

    @Override
    public String toString() {
        return name;
    }
}

File: core/src/main/java/heroes/journey/utils/worldgen/MapGenerationEffect.kt
package heroes.journey.utils.worldgen

import heroes.journey.GameState
import heroes.journey.modlib.Registrable
import heroes.journey.utils.worldgen.MapGenerator.MapGenerationManager
import lombok.Getter

@Getter
abstract class MapGenerationEffect(id: String) : Registrable(id) {

    val dependsOn: MutableList<String> = ArrayList()

    fun register(vararg dependencies: MapGenerationEffect): MapGenerationEffect {
        for (dependency in dependencies) {
            dependsOn.add(dependency.id)
        }
        return register()
    }

    override fun register(): MapGenerationEffect {
        return MapGenerationManager.register(this)
    }

    abstract fun apply(gameState: GameState?)
}
